testapp/exam/code_server.py

```diff
@@ -53,167 +54,311 @@ def timeout_handler(signum, frame):
     """A handler for the ALARM signal."""
     raise TimeoutException('Code took too long to run.')
 
+def create_delete_signal_handler(action=None, old_handler=None):
```

Comment:
Split this into three different functions with clear descriptive names as discussed.